### PR TITLE
WT-13260 Create shared queue mechanism for code coverage

### PR DESCRIPTION
**Pull request by jiechenbo**
### Summary
Create a shared queue mechanism for the parallel work inside code coverage. The advantage is that all created parallel processes would share the load of the tests in the code coverage.

### Detailed Description
Previously the coverage report code would bucket all the tests into their own build directories, and then parallel processes would run them in the bucket. The tests are separated from the top-down and are no specific order. The **change** here is that we create a shared queue mechanism to share the load, with the hopes of equally distributing the load. This will overall improve the coverage report performance. 

Technical details of the code:
- I used a multiprocessing queue that contains all of the build directories. I used the queue as a shared object to be passed into each created process and pass it into an initializer function. The function would then pop off the queue and change their running directory to the build directory.
- Now the run task function will only run one task at a time instead of looping through the task list. Also I changed to use the mechanism executor submit. `executor.submit` pushes the task into it's own queue mechanism inside `concurrent.futures` and therefore covers the shared queue mechanism portion of the ticket.
- I created new analysis mode, which grabs the timing of each task and writes out the tasks in order of timing in descending order into the same json file. The analysis mode is expected to only be run locally for now. The reason is behind the new mode is that it would improve the performance of the shared queue mechanism. Otherwise It is possible that a 
- Per-test code coverage now doesn't have its index build copy as it's name anymore. I pass in the test number and add it into the process, and use that instead of the index.
- I separated out the `task_bucket_info` into `task_list` and `build_dir_list` as it shouldn't be coupled anymore.

### CODE REVIEW

File: test/evergreen/code_coverage/parallel_code_coverage.py [hunk 1]
```diff
@@ -103,6 +94,9 @@ def main():
     if (bucket and bucket != "python" and bucket != "other"):
         sys.exit("Only buckets options \"python\" and \"other\" are allowed")
 
+    if (bucket and analysis):
```
Comment (ajmorton) [inline test/evergreen/code_coverage/parallel_code_coverage.py:97]:
```suggestion
    # Analysis will rewrite the list of coverage tests. If we use this when running a subset of tests only that subset will be written and we'll lose the unscheduled tests.
    if (bucket and analysis):
```

File: test/evergreen/code_coverage/parallel_code_coverage.py [hunk 2]
```diff
@@ -140,14 +135,26 @@ def main():
...
-    logging.debug("task_bucket_info: {}".format(task_bucket_info))
+    logging.debug("task_list: {}".format(task_list))
 
     # Perform task operations in parallel across the build directories
-    run_task_lists_in_parallel(label="tasks", task_bucket_info=task_bucket_info, run_func=run_task_list)
+    analysis_test_timings = run_task_lists_in_parallel(build_dirs_list, task_list, run_func=run_task, analysis=analysis)
+
+    # In analysis mode, we analyze the test and their timings, and sort them in descending order.
+    # A sorted test list will optimize the shared queue across the parallel processes and will
+    # overall shorten the code coverage time
```
Comment (ajmorton) [inline test/evergreen/code_coverage/parallel_code_coverage.py:148]:
```suggestion
    # Running the shortest tests last decreases the amount of time we spend waiting for the last thread to finish the last test, reducing overall runtime
```

**Review by ajmorton — APPROVED** [review]
LGTM with one fix 👍. No need for re-review

File: test/evergreen/code_coverage/parallel_code_coverage.py (hunk 2 above)
Comment (jiechenbo) [reply test/evergreen/code_coverage/parallel_code_coverage.py]: thanks [79008a1](https://github.com/wiredtiger/wiredtiger/pull/10903/commits/79008a1b188ffaeb7af773019d1f9c4c6c7853f3)

File: test/evergreen/code_coverage/parallel_code_coverage.py (hunk 1 above)
Comment (jiechenbo) [reply test/evergreen/code_coverage/parallel_code_coverage.py:99]: thanks [79008a1](https://github.com/wiredtiger/wiredtiger/pull/10903/commits/79008a1b188ffaeb7af773019d1f9c4c6c7853f3)

File: test/evergreen/code_coverage/parallel_code_coverage.py [hunk 3]
```diff
@@ -82,10 +71,12 @@ def main():
                         help='Perform setup actions from the config in each build directory')
     parser.add_argument('-v', '--verbose', action="store_true", help='Be verbose')
     parser.add_argument('-u', '--bucket', type=str, help='Run on only python tests in code coverage')
+    parser.add_argument('-a', '--analysis', action="store_true", help='Analyse and sort the code coverage tests timing')
```
[the remark this answers is not in the document]
Comment (jiechenbo) [reply test/evergreen/code_coverage/parallel_code_coverage.py]: nice suggestion! [79008a1](https://github.com/wiredtiger/wiredtiger/pull/10903/commits/79008a1b188ffaeb7af773019d1f9c4c6c7853f3)

File: test/evergreen/code_coverage/code_coverage_config.json [hunk 4]
```diff
@@ -1,271 +1,255 @@
 {
-  "_comments" : [
+  "_comments": [
     "// Note: As the JSON file format does not support comments it is necessary to use workarounds such",
```
[the remark this answers is not in the document]
Comment (jiechenbo) [reply test/evergreen/code_coverage/code_coverage_config.json:3]: Yes added a comment now [79008a1](https://github.com/wiredtiger/wiredtiger/pull/10903/commits/79008a1b188ffaeb7af773019d1f9c4c6c7853f3)

File: test/evergreen/code_coverage/parallel_code_coverage.py [hunk 5]
```diff
@@ -50,28 +48,19 @@ def run_task_list(task_list_info):
...
-        logging.debug("Finished task {} in {} : took {} seconds".format(task, build_dir, diff.total_seconds()))
-
-    list_end_time = datetime.now()
-    diff = list_end_time - list_start_time
-
-    return_value = "Completed task list in {} : took {} seconds".format(build_dir, diff.total_seconds())
-    logging.debug(return_value)
-
-    return return_value
+    env["GCOV_PREFIX_STRIP"] = build_dir
```
[the remark this answers is not in the document]
Comment (jiechenbo) [reply test/evergreen/code_coverage/parallel_code_coverage.py]: Oops! nice catch [79008a1](https://github.com/wiredtiger/wiredtiger/pull/10903/commits/79008a1b188ffaeb7af773019d1f9c4c6c7853f3)

**Review by jeremythorp — APPROVED** [review]
Thank you - LGTM!

**Pull request merged**